api/eth_types.go

```diff
@@ -0,0 +1,442 @@
+		return err
+	}
+
+	s = strings.Replace(s, "0x", "", -1)
+	if len(s)%2 == 1 {
+		s = "0" + s
+	}
+
+	decoded, err := hex.DecodeString(s)
+	if err != nil {
+		return err
+	}
+
+	*e = decoded
+	return nil
+}
+
+type EthBlock struct {
+	ParentHash       EthHash    `json:"parentHash"`
+	Sha3Uncles       EthHash    `json:"sha3Uncles"`
+	Miner            EthAddress `json:"miner"`
+	StateRoot        EthHash    `json:"stateRoot"`
+	TransactionsRoot EthHash    `json:"transactionsRoot"`
+	ReceiptsRoot     EthHash    `json:"receiptsRoot"`
+	// TODO: include LogsBloom
+	Difficulty    EthUint64 `json:"difficulty"`
+	Number        EthUint64 `json:"number"`
+	GasLimit      EthUint64 `json:"gasLimit"`
+	GasUsed       EthUint64 `json:"gasUsed"`
+	Timestamp     EthUint64 `json:"timestamp"`
+	Extradata     []byte    `json:"extraData"`
+	MixHash       EthHash   `json:"mixHash"`
+	Nonce         EthNonce  `json:"nonce"`
+	BaseFeePerGas EthBigInt `json:"baseFeePerGas"`
+	Size          EthUint64 `json:"size"`
+	// can be []EthTx or []string depending on query params
+	Transactions []interface{} `json:"transactions"`
+	Uncles       []EthHash     `json:"uncles"`
+}
+
+var (
+	EmptyEthHash  = EthHash{}
+	EmptyEthInt   = EthUint64(0)
+	EmptyEthNonce = [8]byte{0, 0, 0, 0, 0, 0, 0, 0}
+)
+
+func NewEthBlock() EthBlock {
+	return EthBlock{
+		Sha3Uncles:       EmptyEthHash,
+		StateRoot:        EmptyEthHash,
+		TransactionsRoot: EmptyEthHash,
+		ReceiptsRoot:     EmptyEthHash,
+		Difficulty:       EmptyEthInt,
+		Extradata:        []byte{},
+		MixHash:          EmptyEthHash,
+		Nonce:            EmptyEthNonce,
+		GasLimit:         EthUint64(build.BlockGasLimit), // TODO we map Ethereum blocks to Filecoin tipsets; this is inconsistent.
+		Uncles:           []EthHash{},
+		Transactions:     []interface{}{},
+	}
+}
+
+type EthCall struct {
+	From     *EthAddress `json:"from"`
+	To       *EthAddress `json:"to"`
+	Gas      EthUint64   `json:"gas"`
+	GasPrice EthBigInt   `json:"gasPrice"`
+	Value    EthBigInt   `json:"value"`
+	Data     EthBytes    `json:"data"`
+}
+
+func (c *EthCall) UnmarshalJSON(b []byte) error {
+	type TempEthCall EthCall
+	var params TempEthCall
+
+	if err := json.Unmarshal(b, &params); err != nil {
+		return err
+	}
+	*c = EthCall(params)
+	return nil
+}
+
+type EthTxReceipt struct {
+	TransactionHash  EthHash     `json:"transactionHash"`
+	TransactionIndex EthUint64   `json:"transactionIndex"`
+	BlockHash        EthHash     `json:"blockHash"`
+	BlockNumber      EthUint64   `json:"blockNumber"`
+	From             EthAddress  `json:"from"`
+	To               *EthAddress `json:"to"`
+	// Logs
+	// LogsBloom
+	StateRoot         EthHash     `json:"root"`
+	Status            EthUint64   `json:"status"`
+	ContractAddress   *EthAddress `json:"contractAddress"`
+	CumulativeGasUsed EthUint64   `json:"cumulativeGasUsed"`
+	GasUsed           EthUint64   `json:"gasUsed"`
+	EffectiveGasPrice EthBigInt   `json:"effectiveGasPrice"`
```

Comment:
Is this a thing? I don't see it in the docs: https://openethereum.github.io/JSONRPC.html#Receipts

Reply:
Those docs are not updated because the OpenEthereum client was deprecated before EIP-1559 kicked in. But this field was introduced then: https://docs.infura.io/infura/networks/ethereum/json-rpc-methods/eth_gettransactionreceipt

Reply:
/ihateeverything